### PR TITLE
Add compression, pickle protocol to comm contexts

### PR DESCRIPTION
**Pull request by mrocklin**
Currently we expect a great degree of uniformity among servers in a Dask
cluster.  This is especially apparent with pickle protocol and
compression formats.  This PR includes a handshake when connecting that
exchanges a bit of information.  Then this information is passed to
serialization functions through the existing context= mechanism.

This allows serialization functions to have more information about the
other side of the connection, and to make choices accordingly.

This isn't yet smooth, but I thought I'd throw it up early so that
people can see what I'm thinking and comment early.

cc @quasiben @jakirkham @jrbourbeau

### CODE REVIEW

File: distributed/comm/tcp.py [hunk 1]
```diff
@@ -552,5 +586,27 @@ class TLSBackend(BaseTCPBackend):
...
+    return {
+        "compression": default_compression,
+        "python": tuple(sys.version_info)[:3],
+    }
+
+
+def handshake_configuration(local, remote):
+    out = {}
+    if local["python"] >= (3, 8) and remote["python"] >= (3, 8):
+        out["pickle_protocol"] = 5
```
Comment (jakirkham) [inline distributed/comm/tcp.py:599]: Would we be able to check whether `pickle5` is available?

File: distributed/protocol/serialize.py [hunk 2]
```diff
@@ -54,11 +54,15 @@ def dask_loads(header, frames):
...
-def pickle_dumps(x):
+def pickle_dumps(x, context=None):
     header = {"serializer": "pickle"}
     frames = [None]
     buffer_callback = lambda f: frames.append(memoryview(f))
-    frames[0] = pickle.dumps(x, buffer_callback=buffer_callback)
+    frames[0] = pickle.dumps(
+        x,
+        buffer_callback=buffer_callback,
+        protocol=context["pickle-protocol"] if context else None,
```
Comment (jakirkham) [inline distributed/protocol/serialize.py:64]: Should this be `_` or `-`. Just noting it is `_` above

File: distributed/protocol/pickle.py [hunk 3]
```diff
@@ -33,15 +33,15 @@ def _always_use_pickle_for(x):
         return False
 
 
-def dumps(x, *, buffer_callback=None):
+def dumps(x, *, buffer_callback=None, protocol=HIGHEST_PROTOCOL):
```
Comment (jakirkham) [inline distributed/protocol/pickle.py:36]: Is this suppose to be `None`? Asking as below we have `protocol or HIGHEST_PROTOCOL`, which seems unneeded if this is `HIGHEST_PROTOCOL`.

**Comment by jakirkham** [comment]
Thanks Matt! 😄 Had a couple minor questions above.

**Comment by mrocklin** [comment]
> Thanks Matt! smile Had a couple minor questions above.

Thanks for the careful review @jakirkham .  In general does this approach seem ok to you?  I'm looking for high level thoughts before I jump in too deeply (there are still some details to iron out before things work smoothly).

File: distributed/comm/tcp.py (hunk 1 above)
Comment (mrocklin) [reply distributed/comm/tcp.py]: Yeah, I was curious about how best to handle `pickle5`.  Good idea.  Done.

File: distributed/protocol/serialize.py (hunk 2 above)
Comment (mrocklin) [reply distributed/protocol/serialize.py]: Thanks, done I think

File: distributed/protocol/pickle.py (hunk 3 above)
Comment (mrocklin) [reply distributed/protocol/pickle.py:36]: Either one is fine with me.  We sometimes pass in `protocol=None`, hence the `or HIGHEST_PROTOCOL` bit, but I figured having the default be informative wouldn't hurt in either case.

**Comment by mrocklin** [comment]
@jcrist the handshake on connect mechanism here may interest you

File: distributed/protocol/pickle.py (hunk 3 above)
Comment (jakirkham) [reply distributed/protocol/pickle.py:36]: Yeah no strong feelings from me either. Just noting it's a little odd to have `HIGHEST_PROTOCOL or HIGHEST_PROTOCOL` potentially ;)

**Comment by jakirkham** [comment]
> > Thanks Matt! smile Had a couple minor questions above.
> 
> Thanks for the careful review @jakirkham . In general does this approach seem ok to you? I'm looking for high level thoughts before I jump in too deeply (there are still some details to iron out before things work smoothly).

Yeah I think so. IIUC this is just the implementation of what was discussed in issue ( https://github.com/dask/distributed/issues/4011 ). So yeah it seems good to me.

Haven't thought too deeply about how the compression side would look. So am trusting you know what you want there ;) Happy to look at that in more detail when that makes sense :)

**Comment by mrocklin** [comment]
> Haven't thought too deeply about how the compression side would look. So am trusting you know what you want there ;) Happy to look at that in more detail when that makes sense :)

Our current policy here isn't great.  It's "if both sides agree on the same default compression then compress, otherwise don't".  In the future we an pass around a set of allowable compressions and then choose the jointly maximal pair.  I figured I'd wait for later to deal with those details though.

**Comment by mrocklin** [comment]
I've just pushed up a commit which moves this from the TCP comm to Comms generally.  Everything seems fine on TLS/Inproc but if someone from RAPIDS can check out UCX I would appreciate that.

**Comment by jakirkham** [comment]
cc-ing @madsbk (in case you have thoughts here 🙂)

File: distributed/comm/tests/test_comms.py [hunk 4]
```diff
@@ -649,7 +646,8 @@ async def handle_comm(comm):
     if os.name != "nt":
         try:
             # See https://serverfault.com/questions/793260/what-does-tlsv1-alert-unknown-ca-mean
-            assert "unknown ca" in str(excinfo.value)
+            # assert "unknown ca" in str(excinfo.value)
+            pass
```
Comment (mrocklin) [inline distributed/comm/tests/test_comms.py:650]: This is a regression.  Our error reporting around bad TLS degraded a bit.  It now shows up as a timeout error with a more generic "could not connect" message.  I tried figuring out what was going on here for a while but couldn't come up with a resolution.  I could use help here.

**Comment by jakirkham** [comment]
> I've just pushed up a commit which moves this from the TCP comm to Comms generally. Everything seems fine on TLS/Inproc but if someone from RAPIDS can check out UCX I would appreciate that.

My guess is this is fine. I'm not sure how much usage we will get out of this in UCX (we ensure `pickle5` is installed as needed and don't currently use compression). That said, it is worth thinking about where we might use this (one idea below).

We probably need to update `to_frames` and `from_frames` to take the `context` in there. Would imagine this is the same for TCP. Though I could always be missing something here.

One thing that might be interesting in the UCX context (and this can be done in another PR) is whether we want to enable/disable `"cuda"` serialization (potentially moving all frames to host) depending on where/how the data is routed. Maybe @pentschev has thoughts on this? 🙂

**Comment by mrocklin** [comment]
Yeah, more generally I encourage RAPIDS folks to think of this PR as two separate features, a specific and a general one

Specific: It's now easier to handle mismatched workers/schedulers with regards to compression/Python versions (maybe not that interesting to you)

General: Comms now have a mechanism to know arbitrary information about both ends of the connection.  If there are questions that you want to know like "which GPU are you?" or "what architecture are you running?" then there is infrastructure here to collect and transfer that information as well as deliver it to any serialization function.

File: distributed/protocol/serialize.py [hunk 5]
```diff
@@ -54,11 +54,15 @@ def dask_loads(header, frames):
...
+def pickle_dumps(x, context=None):
     header = {"serializer": "pickle"}
     frames = [None]
     buffer_callback = lambda f: frames.append(memoryview(f))
-    frames[0] = pickle.dumps(x, buffer_callback=buffer_callback)
+    frames[0] = pickle.dumps(
+        x,
+        buffer_callback=buffer_callback,
+        protocol=context.get("pickle-protocol", None) if context else None,
+    )
```
Comment (jakirkham) [inline distributed/protocol/serialize.py:65]: We will also want to do this [for NumPy `object` arrays]( https://github.com/dask/distributed/blob/6ee7f899d1bc686ba4910c08509c7f2e5345506d/distributed/protocol/numpy.py#L27 ).

**Comment by jakirkham** [comment]
> General: Comms now have a mechanism to know arbitrary information about both ends of the connection. If there are questions that you want to know like "which GPU are you?" or "what architecture are you running?" then there is infrastructure here to collect and transfer that information as well as deliver it to any serialization function.

On this general question, I think there is another point that will come up (though again this would be future work), is can we choose where tasks goes based on this information? For example knowing a worker is "closer" and free might be a useful thing when distributing tasks and moving data, which would be helpful for RAPIDS, but could have broader value. Though there could be other cases like directing tasks based on the capabilities of the worker referenced above.

**Comment by pentschev** [comment]
> One thing that might be interesting in the UCX context (and this can be done in another PR) is whether we want to enable/disable `"cuda"` serialization (potentially moving all frames to host) depending on where/how the data is routed. Maybe @pentschev has thoughts on this? 🙂

I don't really see what how moving all frames to host would benefit UCX. Apart from that, today we only use UCP layer, meaning all routing is done internally by UCX without any capability to expose that to UCX-Py, and consequently to the outside world, such as Dask. That is in itself a limiting factor for doing that.

**Comment by jakirkham** [comment]
> > One thing that might be interesting in the UCX context (and this can be done in another PR) is whether we want to enable/disable `"cuda"` serialization (potentially moving all frames to host) depending on where/how the data is routed. Maybe @pentschev has thoughts on this? 🙂
> 
> I don't really see how moving all frames to host would benefit UCX. Apart from that, today we only use UCP layer, meaning all routing is done internally by UCX without any capability to expose that to UCX-Py, and consequently to the outside world, such as Dask. That is in itself a limiting factor for doing that.

Yeah I wasn't sure if this might be helpful when using IB (especially if we have a more efficient way of doing device-to-host transfers).

**Comment by pentschev** [comment]
> Yeah I wasn't sure if this might be helpful when using IB (especially if we have a more efficient way of doing device-to-host transfers).

For IB this wouldn't be helpful as we rely on GPUDirectRDMA for optimal performance, without that IB performance for CUDA frames isn't that great.

**Comment by mrocklin** [comment]
```python
In [1]: %%time 
   ...: from dask.distributed import Client 
   ...: client = Client("localhost:8786") 
   ...: import dask.array as da 
   ...: x = da.random.random((20000, 20000)) 
   ...: (x + x.T).sum().compute() 
   ...:  
   ...:                                                                                                                                                                                                                                                                                                                                                                                                                                
/home/mrocklin/workspace/distributed/distributed/client.py:1136: VersionMismatchWarning: Mismatched versions found

+---------+---------------+---------------+-------------------------------------+
| Package | client        | scheduler     | workers                             |
+---------+---------------+---------------+-------------------------------------+
| blosc   | 1.7.0         | 1.7.0         | None                                |
| dask    | 2.18.1        | 2.18.1        | 2.20.0                              |
| lz4     | 3.0.2         | 3.0.2         | None                                |
| msgpack | 0.6.2         | 0.6.2         | 1.0.0                               |
| numpy   | 1.17.5        | 1.17.5        | 1.19.1                              |
| python  | 3.7.6.final.0 | 3.7.6.final.0 | {'3.8.5.final.0', '3.6.10.final.0'} |
+---------+---------------+---------------+-------------------------------------+
Notes: 
-  msgpack: Variation is ok, as long as everything is above 0.6
  warnings.warn(version_module.VersionMismatchWarning(msg[0]["warning"]))
CPU times: user 1.38 s, sys: 129 ms, total: 1.51 s
Wall time: 15.4 s
Out[1]: 399987145.9078545
```

**Comment by mrocklin** [comment]
OK, some issues:

1.  I had to turn off pickle5 when sending data to and from the scheduler.  Because the scheduler commonly operates as a proxy between clients and workers it's hard to predict where things will go and the capabilities on the other side.  I think that this probably isn't a *huge* deal because this is rarely where we care about hardcore performance (like worker-worker) and users shouldn't be sending that much data along this wire anyway
2.   I turned off pickle protocol 5 on all of the small serializations like types and such.  This seems more unambiguously ok to me.
3.  We no longer compress msgpack encoded messages.  These are usually small anyway

I'm still fine with these changes, but I thought I'd highlight them in case they make others nervous.

It's pretty cool to see things working in a more heterogenous context though :)

**Comment by jakirkham** [comment]
I'm curious why the Scheduler ends up being a special case. Doesn't it use the same Comms as Workers do? What differs there?

**Comment by mrocklin** [comment]
Because the scheduler commonly operates as a proxy between clients and workers it's hard to predict where things will go and the capabilities on the other side

**Comment by jakirkham** [comment]
Fair enough. At least the usages I see here don't handle out-of-band frames, so would figure using protocol 4 instead of 5 there don't matter much.

The other use case I could see for this functionality would be tracking endianness of workers. Maybe easy to add (though we can defer this as well).

**Comment by mrocklin** [comment]
> The other use case I could see for this functionality would be tracking endianness of workers. Maybe easy to add (though we can defer this as well).

Yeah, I think that this fits under the "general" part of the PR.  We can add whatever we want into the `Comm.handshake_info` method.  That data gets collected from each side and fed into the serialize functions so that we can make well-informed decisions there.  There are probably lots of different use cases. 

As an example someone a long time ago asked for basic token-based security .  That would probably be easy to add with this mechansim.  (cc @Carreau folks you know may be interested)

**Comment by mrocklin** [comment]
The other challenge to this PR is that now that we're sending a couple of messages along at connection time we're uncovering some issues with our stress tests.  So far I don't think that we've identified any issues that we've created, we're just now a bit more sensitive to tests that create and destroy a lot of workers and comms, and so need to be a bit more diligent.  I'm tracking these down, but afterwards I plan to merge.

**Comment by jakirkham** [comment]
SGTM. Thanks Matt! 😄

File: distributed/client.py [hunk 6]
```diff
@@ -2361,7 +2363,10 @@ def get_dataset(self, name, **kwargs):
 
     async def _run_on_scheduler(self, function, *args, wait=True, **kwargs):
         response = await self.scheduler.run_function(
-            function=dumps(function), args=dumps(args), kwargs=dumps(kwargs), wait=wait
+            function=dumps(function, protocol=4),
+            args=dumps(args, protocol=4),
+            kwargs=dumps(kwargs, protocol=4),
```
Comment (jrbourbeau) [inline distributed/client.py:2368]: Why does `protocol=4` need to be specified here (and several other places)?

File: distributed/comm/core.py [hunk 7]
```diff
@@ -175,6 +202,26 @@ async def _():
 
         return _().__await__()
 
+    async def on_connection(self, comm: Comm, handshake_overrides={}):
```
Comment (jrbourbeau) [inline distributed/comm/core.py:205]: Nitpick: I'd feel a little better about initializing `handshake_overrides=None` and then assigning it to be `{}` if `handshake_overrides is None` to avoid any issues with using mutable objects as a keyword default

File: distributed/comm/core.py [hunk 8]
```diff
@@ -187,7 +234,9 @@ def connect(self, address, deserialize=True):
         """
 
 
-async def connect(addr, timeout=None, deserialize=True, **connection_args):
+async def connect(
+    addr, timeout=None, deserialize=True, handshake_overrides={}, **connection_args
```
Comment (jrbourbeau) [inline distributed/comm/core.py:238]: Same comment here about `handshake_overrides`

File: distributed/protocol/compression.py [hunk 9]
```diff
@@ -124,16 +124,22 @@ def zstd_decompress(data):
...
     else:
-        raise ValueError(
-            "Default compression '%s' not found.\n"
-            "Choices include auto, %s"
-            % (default, ", ".join(sorted(map(str, compressions))))
-        )
+        return default_compression
+
+
+get_default_compression()
```
Comment (jrbourbeau) [inline distributed/protocol/compression.py:142]: For my own knowledge, we're just calling `get_default_compression` here to potentially trigger the `ValueError` above?

File: distributed/tests/test_client.py [hunk 10]
```diff
@@ -6137,3 +6137,22 @@ async def test_as_completed_condition_loop(c, s, a, b):
 def test_client_connectionpool_semaphore_loop(s, a, b):
     with Client(s["address"]) as c:
         assert c.rpc.semaphore._loop is c.loop.asyncio_loop
+
+
+@pytest.mark.slow
+@pytest.mark.asyncio
+async def test_mixed_compression(cleanup):
```
Comment (jrbourbeau) [inline distributed/tests/test_client.py:6144]: Nice!

File: distributed/client.py [hunk 11]
```diff
@@ -1310,8 +1310,9 @@ async def _close(self, fast=False):
 
         self.status = "closing"
 
-        for pc in self._periodic_callbacks.values():
-            pc.stop()
+        with suppress(AttributeError):
+            for pc in self._periodic_callbacks.values():
+                pc.stop()
```
Comment (jrbourbeau) [inline distributed/client.py:1315]: Why was this change needed?

File: distributed/client.py [hunk 12]
```diff
@@ -1405,8 +1406,9 @@ def close(self, timeout=no_default):
             return
         self.status = "closing"
 
-        for pc in self._periodic_callbacks.values():
-            pc.stop()
+        with suppress(AttributeError):
+            for pc in self._periodic_callbacks.values():
+                pc.stop()
```
Comment (jrbourbeau) [inline distributed/client.py:1411]: Same question

File: distributed/comm/core.py (hunk 7 above)
Comment (mrocklin) [reply distributed/comm/core.py:205]: In general I agree, but in this case I think that we're safe.

File: distributed/protocol/compression.py (hunk 9 above)
Comment (mrocklin) [reply distributed/protocol/compression.py:142]: Yes.  I'm mostly mimicing the previous behavior in order to not have to think too much about it :)

File: distributed/client.py (hunk 11 above)
Comment (mrocklin) [reply distributed/client.py:1315]: That's a good question.  I'm getting tracebacks like the following, which I don't really understand.  I guess Python can delete an object while it's initializing?  Or maybe this is an at shutdown time thing.

```python
distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:41303
distributed.scheduler - INFO -   dashboard at:           127.0.0.1:36750
Exception ignored in: <function Client.__del__ at 0x7f2d672f2440>
Traceback (most recent call last):
  File "/home/travis/build/dask/distributed/distributed/client.py", line 1189, in __del__
    self.close()
  File "/home/travis/build/dask/distributed/distributed/client.py", line 1413, in close
    if self.asynchronous:
  File "/home/travis/build/dask/distributed/distributed/client.py", line 802, in asynchronous
    return self._asynchronous and self.loop is IOLoop.current()
AttributeError: 'Client' object has no attribute '_asynchronous'
```

Comment (mrocklin) [reply distributed/client.py:1315]: But they're probably not needed.  I can remove them.  I'm just trying a few things to try to placate the stress tests.

**Comment by Carreau** [comment]
> As an example someone a long time ago asked for basic token-based security. That would probably be easy to add with this Mechanism. (cc @Carreau folks you know may be interested)

I can think of  @lheagy, @rcthomas with the Jupyter & Dask deployment at LBL and Nersc (and Shreyas, but I can't find his github handle), And @rpwagner (now at UCSD I believe?).

**Comment by rpwagner** [comment]
> > As an example someone a long time ago asked for basic token-based security. That would probably be easy to add with this Mechanism. (cc @Carreau folks you know may be interested)
> 
> I can think of @lheagy, @rcthomas with the Jupyter & Dask deployment at LBL and Nersc (and Shreyas, but I can't find his github handle), And @rpwagner (now at UCSD I believe?).

Would this be for internal security, like for authenticating communication to and from subtasks, or external security like for public REST API? I'd be able to comment more on the public REST API case.

**Comment by mrocklin** [comment]
Sorry, I was thinking of someone specific that @Carreau and I both know separately.  I didn't mean to bring other folks into this conversation.  I've since followed up with Matthias off-list.

**Comment by rpwagner** [comment]
No problem. If it's something I can contribute to, let me know.

File: distributed/client.py (hunk 11 above)
Comment (jcrist) [reply distributed/client.py:1315]: > I guess Python can delete an object while it's initializing?

Once `__new__` is called and the object instance is created, the `__del__` is always called. So if an error occurs in `__init__` and initialization fails the `__del__` will still be called.

File: distributed/protocol/compression.py (hunk 9 above)
Comment (jrbourbeau) [reply distributed/protocol/compression.py:142]: +1

File: distributed/client.py (hunk 6 above)
Comment (mrocklin) [reply distributed/client.py:2368]: Because these pickle.dumps calls aren't regulated by the same handshake context that exists in the normal serialization process.  We don't know anything about the recipient side here, so we remain conservative.  

In practice I think that this is ok.  These arguments are typically small, and this isn't in high-performance codepaths like direct worker-worker communication during execution.

File: distributed/comm/core.py (hunk 7 above)
Comment (mrocklin) [reply distributed/comm/core.py]: I'm interpretting @jcrist 's :+1: as a vote towards safety.  I've gone ahead and changed these to `None`.

**Comment by mrocklin** [comment]
OK, I've had to relax one test (which is pretty pedantic anyway) and I've resolved all comments.  I'm going to go ahead and merge after passed tests.

Thank you everyone for the careful review.

**Comment by jakirkham** [comment]
Thanks Matt! 😄

**Comment by jakirkham** [comment]
As a result of this, we need to fix the UCX tests as well. Submitted PR ( https://github.com/dask/distributed/pull/4036 ) to do just that.

File: distributed/comm/tests/test_comms.py (hunk 4 above)
Comment (jakirkham) [reply distributed/comm/tests/test_comms.py:650]: Filed as issue ( https://github.com/dask/distributed/issues/4084 ) for tracking and broader visibility.
